stl/src/syncstream.cpp

```diff
@@ -0,0 +1,70 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+// initialize syncstream mutex map
+
+#include <internal_shared.h>
+#include <map>
+#include <shared_mutex>
+
+#pragma warning(disable : 4074)
+#pragma init_seg(compiler)
+
+_STD_BEGIN
+// OBJECT DECLARATIONS
+struct _Mutex_count_pair {
+    shared_mutex _Mutex;
+    uint64_t _Ref_count = 0;
+};
+
+template <class _Ty>
+class _Crt_allocator {
+public:
+    using value_type                             = _Ty;
+    using size_type                              = size_t;
+    using difference_type                        = ptrdiff_t;
+    using propagate_on_container_move_assignment = true_type;
+    using is_always_equal                        = true_type;
+
+    constexpr _Crt_allocator() noexcept = default;
+
+    constexpr _Crt_allocator(const _Crt_allocator&) noexcept = default;
+    template <class _Other>
+    constexpr _Crt_allocator(const _Crt_allocator<_Other>&) noexcept {}
+
+    _NODISCARD __declspec(allocator) _Ty* allocate(_CRT_GUARDOVERFLOW const size_t _Count) {
+        auto _Ptr = _calloc_crt(_Count, sizeof(_Ty));
+        if (!_Ptr) {
+            throw bad_alloc{};
+        }
+        return static_cast<_Ty*>(_Ptr);
+    }
+
+    void deallocate(_Ty* const _Ptr, const size_t) noexcept {
+        _free_crt(_Ptr);
+    }
+};
+
+static map<void*, _Mutex_count_pair, less<void*>, _Crt_allocator<pair<void* const, _Mutex_count_pair>>> _Mutex_map;
+static shared_mutex _Mutex;
+
+extern "C" _NODISCARD _CRTIMP2 shared_mutex* __stdcall _Get_mutex_for_instance(void* _Ptr) noexcept {
+    shared_lock _Guard(_Mutex);
+    auto _Instance_mutex_iter = _Mutex_map.find(_Ptr);
+    _ASSERT_EXPR(_Instance_mutex_iter != _Mutex_map.end(), "No mutex exists for given instance!");
+    return _STD addressof(_Instance_mutex_iter->second._Mutex);

```

Comment:
This is separately compiled (and we control `shared_mutex`) so there's no need to defend against overloaded `operator&`:
```suggestion
    return &_Instance_mutex_iter->second._Mutex;
```